futures-util/src/stream/async_reader.rs

```diff
@@ -0,0 +1,157 @@
+use core::marker::Unpin;
+use core::pin::Pin;
+use futures_core::stream::Stream;
+use futures_core::task::{LocalWaker, Poll};
+use futures_io::AsyncRead;
+use std::cmp;
+use std::io::Result;
+
+/// A stream combinator which interprets a stream of bytes as an `AsyncRead`.
+///
+/// This structure is produced by the `Stream::async_reader` method.
+#[derive(Debug)]
+#[must_use = "streams do nothing unless polled"]
+pub struct AsyncReader<St>
+where
+    St: Stream + Unpin,
+    St::Item: AsRef<[u8]>,
+{
+    stream: St,
+    state: ReadState<St::Item>,
+}
+
+impl<St: Stream + Unpin> Unpin for AsyncReader<St>
+where
+    St: Stream + Unpin,
+    St::Item: AsRef<[u8]>,
+{
+}
+
+#[derive(Debug)]
+enum ReadState<T: AsRef<[u8]>> {
+    Ready { chunk: T, chunk_start: usize },
+    NotReady,
+    Eof,
+}
+
+impl<St> AsyncReader<St>
+where
+    St: Stream + Unpin,
+    St::Item: AsRef<[u8]>,
+{
+    pub(super) fn new(stream: St) -> Self {
+        AsyncReader {
+            stream: stream,
+            state: ReadState::NotReady,
+        }
+    }
+}
+
+impl<St> AsyncRead for AsyncReader<St>
+where
+    St: Stream + Unpin,
+    St::Item: AsRef<[u8]>,
+{
+    fn poll_read(
+        &mut self,
+        lw: &LocalWaker,
+        buf: &mut [u8],
+    ) -> Poll<Result<usize>> {
+        loop {
+            match self.state {
+                ReadState::Ready {
+                    ref mut chunk,
+                    ref mut chunk_start,
+                } => {
+                    let chunk = chunk.as_ref();
+                    let len = cmp::min(buf.len(), chunk.len() - *chunk_start);
+
+                    buf[..len].copy_from_slice(
+                        &chunk[*chunk_start..*chunk_start + len],
+                    );
+                    *chunk_start += len;
+
+                    if chunk.len() == *chunk_start {
+                        self.state = ReadState::NotReady;
+                    }
+
+                    return Poll::Ready(Ok(len));
+                }
+                ReadState::NotReady => {
+                    match Pin::new(&mut self.stream).poll_next(lw) {
```

Comment:
This can use the `ready!` macro to handle the `Poll::Pending` state. (Aside, I'm amazed at how much nicer handling this sort of state machine is with NLL, I'm tempted to update some of my older crates to 2018 just to see the massive readability gain).

Reply:
Oh, also `StreamExt::poll_next_unpin` to avoid the `Pin` construction here (although I think this adapter can be updated to handle `!Unpin` streams pretty easily, but that can be added later).

Reply:
Yeah I was a little shocked this worked. NLL is so wonderful.

Nice tip on `ready!`, I'll switch to that.

Regarding `poll_next_unpin`, that is unfortunately not supported by `TryStream`, which I just switched to. Shall I add it as well?